# **Recursion (Function Calling Itself)**

### Concept Explanation

👉 Recursion = function calling itself

💡 Instead of loop (for, while)
We use function again and again

### Story Time (Mirror Reflection)

Imagine you stand between mirrors 🪞🪞

You see:

- image inside image
- again inside
- again inside

👉 It repeats again and again

💡 That is recursion

### Very Important Rule

👉 Every recursion MUST have a STOP condition

👉 This is called:

👉 Base Case

Without base case ❌ → infinite loop → crash 💥

### Python Simple Example

In [12]:
def countdown(n):
    if n == 0:          # base case
        print("Stop")
    else:
        print(n)
        countdown(n-1)  # function calling itself

In [13]:
countdown(5)

5
4
3
2
1
Stop


### Factorial Example

5! = 5 × 4 × 3 × 2 × 1

In [3]:
def factorial(n):
    if n == 1:              # base case
        return 1
    else:
        return n * factorial(n-1)

In [4]:
print(factorial(5))

120


### Step Thinking

factorial(5)\
= 5 × factorial(4)\
= 5 × 4 × factorial(3)\
= ...\
= 120

### Sum of Numbers

In [6]:
def total(n):
    if n == 0:
        return 0
    else:
        return n + total(n-1)

In [7]:
print(total(5))

15


### Practice Tasks
**Task 1:**

In [16]:
def print_numbers(n):
    if n == 0:
        return
    print(n)
    print_numbers(n - 1)

In [17]:
print_numbers(4)

4
3
2
1


**Task 2**\
def sum_numbers(n):

👉 Return sum of 1 to n

In [ ]:
def sum_numbers(n):
    result = 0
    